crates/transaction-generator-lib/src/lib.rs

```diff
@@ -306,22 +348,121 @@ pub async fn create_txn_generator_creator(
     )
 }
 
-fn get_account_to_burn_from_pool(
-    accounts_pool: &Arc<RwLock<Vec<LocalAccount>>>,
-    needed: usize,
-) -> Vec<LocalAccount> {
-    let mut accounts_pool = accounts_pool.write();
-    let num_in_pool = accounts_pool.len();
-    if num_in_pool < needed {
+/// Simple object pool structure, that you can add and remove from multiple threads.
+/// Taking is done at random positions, but sequentially.
+/// Overflow replaces at random positions as well.
+///
+/// It's efficient to lock the objects for short time - and replace
+/// in place, but its not a concurrent datastructure.
+pub struct ObjectPool<T> {
+    pool: RwLock<Vec<T>>,
+}
+
+impl<T> ObjectPool<T> {
+    pub(crate) fn new_initial(initial: Vec<T>) -> Self {
+        Self {
+            pool: RwLock::new(initial),
+        }
+    }
+
+    pub(crate) fn new() -> Self {
+        Self::new_initial(Vec::new())
+    }
+
+    pub(crate) fn reserve(&self, additional: usize) {
+        self.pool.write().reserve(additional);
+    }
+
+    pub(crate) fn add_to_pool(&self, mut addition: Vec<T>) {
+        assert!(!addition.is_empty());
+        let mut current = self.pool.write();
+        current.append(&mut addition);
         sample!(
-            SampleRate::Duration(Duration::from_secs(10)),
-            warn!("Cannot fetch enough accounts from pool, left in pool {}, needed {}", num_in_pool, needed);
+            SampleRate::Duration(Duration::from_secs(120)),
+            info!("Pool working set increased to {}", current.len())
         );
-        return Vec::new();
     }
-    accounts_pool
-        .drain((num_in_pool - needed)..)
-        .collect::<Vec<_>>()
+
+    pub(crate) fn add_to_pool_bounded(
+        &self,
+        mut addition: Vec<T>,
+        max_size: usize,
+        rng: &mut StdRng,
+    ) {
+        assert!(!addition.is_empty());
```

Comment:
Either no-op or truncate to max_size instead of assert ?

Reply:
again, it's there for easier catching of bugs. if it becomes bothersome , we can remove.